Settings filter duplicates

rakons opened the pull request:
This PR adds an option to filter duplicates when direct loading data from file and fcb backend.
The average complexity should be ```N ^ sqrt(2)```, thanks to the fact that not all the data is read on every entry, but only all the data left.

rakons committed:
settings: fcb: Duplicates filtering option

This commit adds a possibility to activate duplicates filtering
during direct loading.

JIRA: NCSDK-3017

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

rakons committed:
settings: file: Duplicates filtering option

This commit adds a possibility to activate duplicates filtering
during direct loading.

JIRA: NCSDK-3017

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

Laczen commented:
@rakons, what do you mean by filter duplicates ? Is the idea to call the settings_handler just for the latest entry in the fs, or is it something else ?

rakons commented:
@Laczen - yes. Ignore previous (old) copies and call the cb function only for the final data.

Laczen commented:
@rakons, OK. In that case I think you can achieve the same results without any changes to fcb. What you are trying to achieve is similar to what is done in the compress method. In that method a copy of an entry is done when there is no later entry in the fs that has the same name. In your case you would only call the load if there is no later entry in the fs with the same name. The load can be modified very easily to do the same functionality as the compress, you can remove the fcb_walk routine completely. For the nffs case the same methodology can be used. I would propose the filter setting to be a configuration option. By introducing a load variable that is set to true at the start for every entry and then walking over the remaining fs to see if you need to set load to false (if you find another entry with the same name) it is easy to use the configuration option to decide if you need to walk over the remaining items or not.
At a high level it would look like:
```
while (fcb_getnext(&cf->cf_fcb, &loc1.loc) == 0) {
   read name1
   load = true;
#if enabled(SETTINGS_SFCB_FILTER_DUPLICATE)
   loc2 = loc1;
   while (fcb_getnext(&cf->cf_fcb, &loc2.loc) == 0) {
       read name2
       if name1 equal to name2 {
           load = false;
           break;
       }
    }
#endif
   if (load) {
      call settings_handler
   }
}
```

rakons commented:
It makes a sense. Limiting callbacks levels inside settings is always good move. Shouldnt we lock fcb while processing it using fcb_getnext?

Laczen commented:
> It makes a sense. Limiting callbacks levels inside settings is always good move. Shouldnt we lock fcb while processing it using fcb_getnext?

@nvlsianpu, ^^^ should fcb be locked ?

rakons commented:
Reworked limiting callbacks level by @Laczen sugestion.
Currently fcb is not locked during data traversing. Waiting for @nvlsianpu to feedback if it should be locked. It seems that we are missing some locking while data compressing anyway inside fcb.

nvlsianpu commented:
>> It makes a sense. Limiting callbacks levels inside settings is always good move. Shouldnt we lock fcb while processing it using fcb_getnext?

> @nvlsianpu, ^^^ should fcb be locked ?

I have been Looking deeply into code. Answer to this question is yes. Need to introduce few more lock into code for ensure multiple access compatibility

I have been Looking deeply into code. Answer to this question is yes. Need to introduce few more lock into code for ensure multiple access compatibility. I Discussed this with @rako of-channel.

Laczen commented:
> > > It makes a sense. Limiting callbacks levels inside settings is always good move. Shouldnt we lock fcb while processing it using fcb_getnext?
> 
> > @nvlsianpu, ^^^ should fcb be locked ?
> 
> I have been Looking deeply into code. Answer to this question is yes. Need to introduce few more lock into code for ensure multiple access compatibility
> 
> I have been Looking deeply into code. Answer to this question is yes. Need to introduce few more lock into code for ensure multiple access compatibility. I Discussed this with @rako of-channel.

@rakons, @nvlsianpu, are you sure about the locking ? AFAIR, locking in fcb is only required when doing a write (nothing else changes the fs), when a writing is done fcb allocates the space and locks the fs (fcb_append), data is written during this lock (first unlocking it and then doing the write, afterwards lock it back again). when the writing is finished (fcb_close) the last part is written and then the fs is unlocked.

Laczen commented on include/settings/settings.h:
```diff
@@ -268,16 +268,18 @@ typedef int (*settings_load_direct_cb)(
+ * @param[in]     cb                pointer to the callback function.
+ * @param[in,out] param             parameter to be passed when callback
+ *                                  function is called.
+ * @param[in]     filter_duplicates filter all duplicated entry.
```
The proposed addition IMO does not justify an API change. Either enable it by default or use the filter_duplicate option only on settings_load_direct().

rakons replied:
It does not make a big sense to use it when all the settings are loaded at startup. It was the idea - do not change an external API. If somebody knows it is here and wishes to use it - set the specific field in the settings.

Laczen replied:
You are changing an external API. Settings can be configured to use a custom backend, for these you are changing the API.

rakons replied:
@nvlsianpu - What is your opinion? Anybody more we can vote if it is worth to change API between settings and backend or is is better to add another backend function?
The first solution would fail during compilation if not adjusted. The second can lead to runtime error but if the functionality is not used, it would not require any adjustment from the user.

Laczen replied:
@rakons, there are other options: enable it by default (makes the load a little longer) or use config to enable it.

nvlsianpu replied:
I would introduce this API change - it is no big change. Making it optional makes things harder to be maintained - I believe adaptation of a current custom back-end will be easy.

Laczen replied:
@nvlsianpu, why not enable it by default ?

nvlsianpu replied:
This API was introduced like ~2 week ago by https://github.com/zephyrproject-rtos/zephyr/pull/19077

Laczen replied:
@nvlsianpu ?? There is a new change used for passing the bool filter_duplicates.

Laczen commented on include/settings/settings.h:
```diff
@@ -268,16 +268,18 @@ typedef int (*settings_load_direct_cb)(
 	settings_load_direct_cb cb,
-	void                   *param);
+	void                   *param,
+	bool                    filter_duplicates);
```
See above

Laczen commented on include/settings/settings.h:
```diff
@@ -375,6 +377,15 @@ struct settings_load_arg {
+	 * The backends that does not load old settings by design may ignore
+	 * this flag.
+	 */
+	bool filter_duplicates;
```
See above

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,102 @@ int settings_fcb_dst(struct settings_fcb *cf)
+		if (rc) {
+			return rc;
+		}
+		if (!strcmp(name, name2))
```
strcmp is not sufficient, suppose name ="bt/device" and name2="bt/device1"

rakons replied:
strcmp("bt/device", "bt/device1") != 0 - conditional part not executed. What's the problem?

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,102 @@ int settings_fcb_dst(struct settings_fcb *cf)
+
+		if (filter_duplicates) {
+			rc = settings_fcb_check_duplicate(cf, &entry_ctx, name);
+			if (rc < 0)
```
shouldn't this be if (rc <= 0) ?

rakons replied:
No. If 0 is returned I wish to set pass_entry to 1 and continue the loop.

Laczen replied:
There is something strange with the logic here, if no duplicates are found or if a duplicate is found you want to set pass_entry to 1 (unless the description of settings_fcb_check_duplicate is wrong).

rakons replied:
```pass_entry = (rc == 0);``` - bracket added. If duplicates are found then rc = 1. If no duplicates is found - rc = 0.

Laczen replied:
Ok, this makes it more clear, but I still think it is not correct. Whenever an error is occurring during the find duplicates there is a jump out of the loop, while it should just be a continue. It should still load whatever it has found.

rakons replied:
Done

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -81,23 +84,65 @@ static int settings_file_load_priv(struct settings_store *cs, line_load_cb cb,
+				next_buf[next_len] = '\0';
+
+				if (!strcmp(next_buf, buf)) {
+					is_duplicate = true;
```
strcmp is not sufficient, see above.

rakons replied:
I believe it is - see above.

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -81,23 +84,65 @@ static int settings_file_load_priv(struct settings_store *cs, line_load_cb cb,
 	}
 
 	while (1) {
+		bool is_duplicate = false;
```
You could easily use the exact same methodology in fcb, doing the same thing makes it easier to maintain.

rakons replied:
I would rather pass the same variable name into file - is_duplicate does not explain everything and I am worried that it might be used for additional checking in the future. pass_entry explains what this variable does.

rakons replied:
Done

Laczen replied:
I agree, I would not use is_duplicate, but use a boolean load = true. This explains everything, you set load to false when not needed. The methodology in the file is a lot simpler than in the fcb backend, so keep this methodology.

rakons replied:
I do not think so. Moreover - to much of tabs is required in files and it makes it hard to keep line length limit. Splitting it into two functions line in fcb backend.

Laczen replied:
It is very easy to put this in a function for the file as well. This is no reason. Keep the methods equal! Just have this function return 0 if it finds a match, and an error code when it doesn't (a error: end of file or whatever error means loading is needed).

rakons replied:
I see no reason to flatten the function return value so much. It might make sense to react in the same way when an error occur and no match is found. The functionality is the same now in both backends.

Laczen replied:
@rakons, I do see a reason for both backends to flatten out the function return. At the moment settings reads as much as it can, when you don't flatten out the error settings will return no value at all because there is a error later in the backend, even when this error has nothing to do with the value being read.

rakons replied:
Did you look into the current code?

nvlsianpu commented:
@Laczen fcb settings compression: when one thread is appending to scratch, other thread might append to, so it might occur that it causes even scratch sector full filling before compression is done.

What i see is that the fcb instance is preserved against its state violation., but not against something like that.

Laczen commented:
> @Laczen fcb settings compression: when one thread is appending to scratch, other thread might append to, so it might occur that it causes even scratch sector full filling before compression is done.
> What i see is that the fcb instance is preserved against its state violation., but not against something like that.
True, but fcb cannot protect against this, I think the settings lock protects against this. But of course this means you can only use the fcb fs for settings.

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -49,19 +53,62 @@ int settings_file_dst(struct settings_file *cf)
+
+		rc = settings_next_line_ctx(&entry2_ctx);
+		if (rc || entry2_ctx.len == 0) {
+			break;
```
continue ?

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -49,19 +53,62 @@ int settings_file_dst(struct settings_file *cf)
+			return rc;
+		}
+		if (name2_len == 0) {
+			return -ENODATA;
```
continue ?

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -49,19 +53,62 @@ int settings_file_dst(struct settings_file *cf)
+	/* Searching the duplicates */
+	while (1) {
+		char name2[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name2_len;
```
Move to beginning of routine ?

rakons replied:
Why?

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -81,23 +128,34 @@ static int settings_file_load_priv(struct settings_store *cs, line_load_cb cb,
 	}
 
 	while (1) {
```
while (settings_next_line_ctx(&entry_ctx) == 0) ?

rakons replied:
Ok.

Laczen commented on subsys/settings/src/settings_file.c:
```diff
@@ -81,23 +128,34 @@ static int settings_file_load_priv(struct settings_store *cs, line_load_cb cb,
 	while (1) {
-		rc = settings_next_line_ctx(&entry_ctx);
+		char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name_len;
```
Move to beginning of routine ?

rakons replied:
Any good reason?

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-	argp = (struct settings_fcb_load_cb_arg *)arg;
-
-	size_t len_read;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry2_ctx.loc)) == 0) {
```
remove rc, it is not used.

rakons replied:
done

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-	size_t len_read;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry2_ctx.loc)) == 0) {
+		char name2[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name2_len;
```
Move to the start of the routine, I know its not really required, its just how most routines are written in zephyr.

rakons replied:
This is totally valid C code and I do not see any reason to move variable declaration outside the scope where it is really used.

rakons replied:
Rather wont do, unless forced by more people.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-	return 0;
+	if (rc == -ENOTSUP) {
+		rc = 0;
+	}
```
remove

rakons replied:
Done

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
+					     &entry2_ctx);
+		name2[name2_len] = '\0';
+		if (rc) {
+			return rc;
```
continue ? and move the if before name2[...]=. Whenever something is found in fcb that is not a settings item it should just be ignored.

rakons replied:
(rc) removed, but break on error - continuing makes the code enter into endless loop.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
+	if (rc == -ENOTSUP) {
+		rc = 0;
+	}
+	return rc;
```
return 0

rakons replied:
return false, returned type updated to bool.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-	if (rc) {
-		return -EINVAL;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry_ctx.loc)) == 0) {
+		char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
```
remove rc, it is not used

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-	rc = fcb_walk(&cf->cf_fcb, 0, settings_fcb_load_cb, &arg);
-	if (rc) {
-		return -EINVAL;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry_ctx.loc)) == 0) {
```
remove rc, it is not used

rakons replied:
Let it be your way. Do not really care but I really do not like to loss the error information too early.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
-		return -EINVAL;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry_ctx.loc)) == 0) {
+		char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name_len;
```
Move to the start of the routine

rakons replied:
This is totally valid C code and I do not see any reason to move variable declaration outside the scope where it is really used.

rakons replied:
Rather wont do, unless forced by more people.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry_ctx.loc)) == 0) {
+		char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name_len;
+		int rc;
```
remove

rakons replied:
Ok

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
+		if (rc) {
+			LOG_ERR("Failed to load line name: %d", rc);
+			return 0;
+		}
```
whenever something is found that does not belong to settings, just continue

rakons replied:
Tried that - endless loop detected during tests.

Laczen replied:
That is strange, I think this means the loop is not ended by `fcb_getnext()` but by the `settings_line_name_read()`, maybe @nvlsianpu has an idea why this happens.

rakons replied:
Ok... tested it again after code changes - it works. The problem is only in file backend. 
```
f (entry2_ctx.len == 0) {
			continue;
		}
```
makes it looping forever

rakons replied:
Added 'continue' where it works during unittests.

Laczen replied:
OK, this is normal because this is to get out of the `while (1)` at the end of the fs. The while(1) could be replaced to get complete similar code for fcb and fs, but I see no reason to push it this far. **NICE WORK**

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +75,97 @@ int settings_fcb_dst(struct settings_fcb *cf)
+		}
+	}
+	if (rc == -ENOTSUP) {
+		rc = 0;
```
remove as it is not used

rakons replied:
Yep. Sure. Now it is not.

Laczen commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +76,92 @@ int settings_fcb_dst(struct settings_fcb *cf)
-		return 0;
+		if (settings_line_name_read(name2, sizeof(name2), &name2_len,
+					    &entry2_ctx)) {
+			break;
```
```
LOG_ERR("failed to load line...");
continue;
```

rakons replied:
Done

Laczen commented:
For me this PR is ready for merging, only thing to be decided: change the API as proposed or just enable the loading of the latest element (filter_duplicates = true) as default or as a configuration option.

IMO enable it as default would be the best option.

@rakons, **THANKS**.

nvlsianpu commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +76,93 @@ int settings_fcb_dst(struct settings_fcb *cf)
+		cs,
+		settings_line_load_cb,
+		(void *)arg,
+		arg && arg->filter_duplicates);
```
does arg can be null now?

rakons replied:
Yes

rakons commented:
The NVS unit test requires update.

rakons commented:
Test updated for nvs backend (or any other that may not preserve the order of the written data).

nvlsianpu commented:
@rakons Need fix test building issues.

rakons commented:
@nvlsianpu Everything is green now :)

nvlsianpu commented:
^^szacun

Laczen reviewed (changes requested):
I am still waiting for a good reason to change the API, I know the change is small but IMO nobody will use settings in a mixed case with linear loading at start and loading the newest value in other cases and only when requested by the loading request. The influence on the settings_load time will be small and it will be much simpler if all settings_load calls just return the newest value. Also it will have a positive influence on code size.

carlescufi commented:
> I am still waiting for a good reason to change the API, I know the change is small but IMO nobody will use settings in a mixed case with linear loading at start and loading the newest value in other cases and only when requested by the loading request. The influence on the settings_load time will be small and it will be much simpler if all settings_load calls just return the newest value. Also it will have a positive influence on code size.

I think the change is so trivial that it can be introduced safely, but I am not an expert on settings so I will ask others. 

@fnde-ot @jhedberg @Vudentz could you please voice your opinion about [this particular conversation](https://github.com/zephyrproject-rtos/zephyr/pull/19541/files#r331561672)

fnde-ot reviewed (approved):
Maybe we should consider deprecating backends storing duplicates at some point, but meanwhile this change makes sense to me. Nice job!

rakons commented:
The output for the meeting with @carlescufi and @jhedberg:
We are moving towards the backends that does not present the entity history.
FCB stays here for backward compatibility, FILE would always be there, but we does not expect performance from FILE backend.
What we should do:

- [ ] Make an doc update that from 2.1 we does support the backends that presents the history (then we may remove some complexity from the modules that uses settings).
- [ ] Filter out deleted entries.
- [ ] Add filtering also for global loading function.

rakons committed:
settings: doc: Update the backend documentation

This commit updates the settings backend documentation
to clearly state that it cannot provide old entities
before the final one.

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented:
Seems done
@carlescufi - who can check the documentation?

carlescufi commented:
> Seems done
> @carlescufi - who can check the documentation?

@dbkinder will, he is automatically added when you edit `.rst` files.

carlescufi commented:
@Laczen could you take another look?

jhedberg commented on subsys/settings/src/settings_fcb.c:
```diff
@@ -79,52 +76,102 @@ int settings_fcb_dst(struct settings_fcb *cf)
+static int settings_fcb_load_priv(struct settings_store *cs,
+				  line_load_cb cb,
+				  void *cb_arg,
+				  bool filter_duplicates)
```
What's the reason you left the `filter_duplicates` as a parameter for this internal API?

rakons replied:
It is used to search the duplicates when storing. Storing procedure can have simpler algorithm to remove the duplicates with a * (x) complexity. It is left here for storing performance.

jhedberg replied:
@rakons ok, makes sense, thanks for the clarification!

dbkinder commented on doc/reference/storage/settings/settings.rst:
```diff
@@ -94,13 +94,17 @@ Loading data from persisted storage
 retrieved.
 
+Technically FCB and filesystem backends may store some history of the entities.
+It means that old data are storied first and then new data entity is stored
```
For clarity and fixing some misspellings, how about instead:
```
This means that the newest data entity is stored after any older existing data entities.
Starting with Zephyr 2.1, the back-end must filter out all old entities and call the
callback with only the newest entity.
```

rakons replied:
Fixed. Please check if this is what you meant.

rakons commented on doc/reference/storage/settings/settings.rst:
```diff
@@ -94,13 +94,17 @@ Loading data from persisted storage
 retrieved.
 
+Technically FCB and filesystem backends may store some history of the entities.
+It means that old data are storied first and then new data entity is stored
```
Covered by @dbkinder comment below.

carlescufi commented:
@rakons can you fix the checkpatch issues?
